src/clusterfuzz/_internal/bot/fuzzers/centipede/engine.py

```diff
@@ -0,0 +1,221 @@
+
+    fuzz_result = runner.run_and_wait(
+        additional_args=arguments, timeout=max_time + _CLEAN_EXIT_SECS)
+    log_lines = fuzz_result.output.splitlines()
+    fuzz_logs = '\n'.join(log_lines)
+
+    crashes = []
+    # Stats report is not available in Centipede yet.
+    stats = None
+    for line in log_lines:
+      reproducer_path = _get_reproducer_path(line)
+      if reproducer_path:
+        crashes.append(
+            engine.Crash(reproducer_path, fuzz_logs, [],
+                         int(fuzz_result.time_executed)))
+        continue
+    return engine.FuzzResult(fuzz_result.output, fuzz_result.command, crashes,
+                             stats, fuzz_result.time_executed)
+
+  def reproduce(self, target_path, input_path, arguments, max_time):  # pylint: disable=unused-argument
+    """Reproduce a crash given an input.
+
+    Args:
+      target_path: Path to the target.
+      input_path: Path to the reproducer input.
+      arguments: Additional arguments needed for reproduction.
+      max_time: Maximum allowed time for the reproduction.
+
+    Returns:
+      A ReproduceResult.
+    """
+    runner = new_process.UnicodeProcessRunner(target_path, [input_path])
+    result = runner.run_and_wait(timeout=max_time)
+
+    return engine.ReproduceResult(result.command, result.return_code,
+                                  result.time_executed, result.output)
+
+  def _create_temp_dir(self, name):
+    """Creates temporary directory for fuzzing."""
+    new_directory = os.path.join(fuzzer_utils.get_temp_dir(), name)
+    engine_common.recreate_directory(new_directory)
+    return new_directory
+
+  def minimize_corpus(self, target_path, arguments, input_dirs, output_dir,
+                      reproducers_dir, max_time):
+    """Optional: run corpus minimization.
+    Args:
+      target_path: Path to the target.
+      arguments: Additional arguments needed for corpus minimization.
+      input_dirs: Input corpora.
+      output_dir: Output directory to place minimized corpus.
+      reproducers_dir: The directory to put reproducers in when crashes are
+          found.
+      max_time: Maximum allowed time for the minimization.
+
+    Returns:
+      A FuzzResult object.
+    """
+    raise NotImplementedError
+
+  def minimize_testcase(self, target_path, arguments, input_path, output_path,
+                        max_time):
+    """Optional: Minimize a testcase.
+    Args:
+      target_path: Path to the target.
+      arguments: Additional arguments needed for testcase minimization.
+      input_path: Path to the reproducer input.
+      output_path: Path to the minimized output.
+      max_time: Maximum allowed time for the minimization.
+    Returns:
+      A ReproduceResult.
+    Raises:
+      TimeoutError: If the testcase minimization exceeds max_time.
+    """
+    raise NotImplementedError
+
+  def cleanse(self, target_path, arguments, input_path, output_path, max_time):
+    """Optional: Cleanse a testcase.
```

Comment:
I think you forgot to get rid of "Optional", which I don't think makes sense here.

Reply:
Done this time, and other `os.chmod()`s.